pandas/core/computation/parsing.py

```diff
@@ -0,0 +1,154 @@
+""":func:`~pandas.eval` source string parsing functions
+"""
+
+from io import StringIO
+from keyword import iskeyword
+import token
+import tokenize
+from typing import Iterator, Tuple, Union
+
+# A token value Python's tokenizer probably will never use.
+BACKTICK_QUOTED_STRING = 100
+
+Tok = Tuple[int, str]
+
+
+def create_valid_python_identifier(name: str) -> str:
+    """Create valid Python identifiers from any string.
+
+    Check if name contains any special characters. If it contains any
+    special characters, the special characters will be replaced by a special
+    string and a prefix is added.
+
+    Raises
+    ------
+    If the returned name is not a Python valid identifier, raise an exception.
+    This can happen if there is a hashtag in the name, as the tokenizer will
+    than terminate and not find the backtick.
+    But also for characters that fall out of the range of (U+0001..U+007F).
+
+    """
+    if name.isidentifier() and not iskeyword(name):
+        return name
+
+    # Create a dict with the special characters and their replacement string.
+    # EXACT_TOKEN_TYPES contains these special characters
+    # toke.tok_name contains a readable description of the replacement string.
+    special_characters_replacements = {
+        char: "_{}_".format(token.tok_name[tokval])
+        # The ignore here is because of a bug in mypy that is resolved in 0.740
+        for char, tokval in tokenize.EXACT_TOKEN_TYPES.items()  # type: ignore
+    }
+    special_characters_replacements.update(
+        {
+            " ": "_",
+            "?": "_QUESTIONMARK_",
+            "!": "_EXCLAMATIONMARK_",
+            "$": "_DOLLARSIGN_",
+            "€": "_EUROSIGN_",
+            # Including quotes works, but they have to come in pairs.
+            "'": "_SINGLEQUOTE_",
+            '"': "_DOUBLEQUOTE_",
+            # Currently not possible. Terminates parser and won't find backtick.
+            # "#": "_HASH_",
+        }
+    )
+
+    name = "".join(special_characters_replacements.get(char, char) for char in name)
+    name = "BACKTICK_QUOTED_STRING_" + name
+
+    if not name.isidentifier():
+        raise ValueError(
+            "Could not convert {} to a valid Python identifier.".format(name)
+        )
+
+    return name
+
+
+def tokenize_and_clean_column_names(name: Union[str, int]) -> Union[str, int]:
```

Comment:
why would this take/return an int?

Reply:
This function is called for all column names regardless of they are being used in the query. For column names, pandas will accept integers as well. That's why this function must also accept integers, so that for the following example, query works as well. 

```python
import pandas as pd
df = pd.DataFrame({0: [1, 2, 3], "A A": [2, 3, 4]})
df.query("`A A` > 3")
```

Reply:
As a name resolver with integers is worthless, I removed the inter names before being inserted into the dictionary. This will simplify some type hints, and fallback code for integer names.